# Local Ancestry Inference methods pipeline

This is a pipeline to detect introgression with Local Ancetry Inference through various methods: Loter, RFMix, ELAI and Recomb-Mix


# 1. Loter (Dias-Alves et al. 2018 MBE)
```
# First, we extract the datasets NEGW (30 individuals), IBGW (32 inds.) and DOG (31 inds.)

in=/path_to_phased_pops
out=/path_to_loter_output

loter_cli --ref $in/negw.phased.vcf.gz $in/dog.phased.vcf.gz --adm $in/ibgw.phased.vcf.gz --verbose --format vcf --ncore 30 --output $out/ibgw.adm.txt
```

# 2. RFMix (Maples et al. 2013 Am J H Gen)

We followed MartÃ­n's pipeline at https://github.com/armartin/ancestry_pipeline, using RFMixv1.5.4
First of all, we need a calculation of current population effective size (Ne), done with GONE (Santiago et al. 2020, MBE)

## Pre-step: calculate Ne with GONE
```
# GONE uses .map and .ped files to run, so we will run PLINK on the IBGW dataset. 

in=/path_to_phased_pops
out=/path_to_RFMix_input

plink --vcf $in/ibgw.phased.vcf.gz --recode --make-bed --out $out/autosomes.ibgw --dog

# Then, we modified the INPUT_PARAMETERS_FILE that comes with the GONE program. 

cat $out/INPUT_PARAMETERS_FILE 
#INPUT_PARAMETERS_FILE
PHASE=1 ### Phase = 0 (pseudohaploids), 1 (known phase), 2 (unknown phase)
cMMb=1  ### CentiMorgans per Megabase (if distance is not available).
DIST=1  ### none (0), Haldane correction (1) or Kosambi correction (2)
NGEN=2000 ### Number of generations for which linkage data is obtained in bins
NBIN=400  ### Number of bins (e.g. 1000, so that each bin includes NGEN/NBIN generations)
MAF=0.001   ### Minimum allele frequency (0-1)
ZERO=1    ### 0: Remove SNPs with zeroes (1: allow for them)
maxNCHROM=-99  ### Maximum number of chromosomes to be analysed (-99 = all chromosomes; maximum number is 200)
maxNSNP=100000 ### Maximum approx number of SNPs per chromosomes to be analysed (max is 100000)
hc=0.05   ### Maximum value of c analysed
REPS=40   ### Number of replicates to RUN GONE
threads=-99  ### Number of threads (if -99 it uses all possible processors)

# To run GONE, we just use script_GONE.sh, which is a script given with the program. 

bash $out/script_GONE.sh $out/autosomes.ibgw

# Once Ne is calculated with GONE, we can run RFMix. We will use a physical map of domestic dogs (Auton et al. 2013, Plos Gen).
```
## step a. Running SHAPEIT2 (Delaneau et al. 2013, Nat Meth). It needs to be run per chromosome. 
```
shap=/path_to_shapeit/shapeit
map=/path_to_maps

$shap --input-vcf $out/chr01.ibgw.phased.recode.vcf \
	--input-map $map/Auton_map.chr01.shapeit.txt \
	--seed 500 \
	--window 0.5 \
	--effective-size 210 \ # calculated with GONE
	-O $file/chr01.ibgw.phased.shapeit \
	--force --thread 1
```

# step b. Running shapeit2rfmix.py from Martin's pipeline 
```
prog=/path_to_Martin_pipeline

cat dog.list negw.list > ref.list # list of source individuals (NEGW + DOG)

for i in {01..38}; do echo chr$i
python2.7 $prog/shapeit2rfmix.py \
--shapeit_hap_ref $out/chr$i.ibgw.phased.shapeit.haps \
--shapeit_hap_admixed $out/chr$i.ibgw.phased.shapeit.haps \
--shapeit_sample_ref $out/ibgw.phased.sample \
--shapeit_sample_admixed $out/ibgw.phased.sample \
--ref_keep $out/ref.list \
--admixed_keep $out/ibgw.list \
--chr ${i} \
--genetic_map $maps/Auton_map.chr$i.shapeit.txt \
--out $out/chr$i.ibgw
done
```
# step c. Running RFMix from Martin's pipeline 
```
input=/path_to_RFMix_input
rfmix=/path_to_rfmix/RFMix_v1.5.4

for i in {01..38}; do
python2.7 $rfmix/RunRFMix.py \
-e 2 -w 0.001 -G 13 --num-threads 14 \ # with G, we select the number of generations: 13, 100, 1000.
--use-reference-panels-in-EM --forward-backward \ # -w will be used to run different window sizes; 1kb: -w 0.001 & 100kb: -w 0.1)
PopPhased $input/chr$i.ibgw_all_chr$i.alleles $input/ibgw.dog.negw.classes $input/chr$i.ibgw_all_chr$i.snp_locations \
-o $input/chr$i.100kb.13gen.rfmix
done
```
# step d. Running a modified version of collapse_ancestry.py (Martin's pipeline) to accomodate 38 chromosomes
```
# we run this command per each genome. In this case, genome=CluJAL7494
python2.7 $prog/collapse_ancestry.dog.py --rfmix $in/chr1.1kb.13gen.rfmix.2.Viterbi.txt --snp_locations $in/chr1.ibgw.snp_locations --fbk $in/chr1.1kb.13gen.rfmix.2.ForwardBackward.txt --fbk_threshold 0.99 --ind $genome --ind_info $in/ibgw.dog.negw.sample --pop_labels NEGW,DOG,UNK --out $genome.13gen.1kb
```


# 3. ELAI (Guan et al. 2014 Genetics)

# step a. Per each chromosome in each population, We generate a .bimbam file
```
for i in dog ibgw negw; do echo $i
for j in {01..38}; do
vcftools --gzvcf $i.phased.vcf.gz --chr chr$j --recode --out chr$j.$i
awk -F'\t' '!/^#/ { $3 = "rs" $2 }1' OFS='\t' chr$j.$i.recode.vcf > chr$j.$i.rs.recode.vcf
plink --vcf chr$j.$i.rs.recode.vcf --recode bimbam --dog --out chr$j.$i.rs.bimbam
done; done

```
# step b. We run ELAI using the .bimbam files

```
in=/path_to_ELAI_output
elai=/path_to_ELAI_software/elai-lin

for j in {01..38}; do
$elai -g $in/chr$j.negw.rs.bimbam.recode.geno.txt -p 10 -g $in/chr$j.dog.rs.bimbam.recode.geno.txt -p 11 -g $in/chr$j.ibgw.rs.bimbam.recode.geno.txt -p 1 -pos $in/chr$j.ibgw.rs.bimbam.recode.pos.txt -s 20 -C 2 -c 10  -mixgen 13 --exclude-nopos --exclude-miss1 -o chr$j
done

# we choose the number of generations ago when they mixed with -mixgen (13, 100, 1000)
```


# 4. Recomb-Mix (Wei & Zhang 2023 bioRxiv - https://github.com/ucfcbb/Recomb-Mix)

```
rec=/path_to_Recomb-Mix/RecombMix_v0.4 
in=/path_to_intput
out=/path_to_output
reco=/home/carlos/Desktop/programs/dog_recombination

# we will use the DOG and NEGW source populations as a reference panel population. 

for i in {01..38}; do echo chr$i
$rec -p $in/dog.negw.chr$i.phased.vcf.gz -q ibgw.chr$i.phased.vcf.gz -a ref_panel_pop_label.txt -o $out -u 1
done


