# Detection of putative alleles under selection 

We want to detect which alleles are putatively under selection in IBGW, but not in NEGW. To do this, we run two cross population statistics (XPnSL and XPEHH) with selscan v2.0 (Szpiech 2024, Bioinformatics)
Even though selscan v2.0 can work with unphased data, to increase accuracy we are using phased data. 
We extracted the chromosomes and positions from every chromosomal VCF and output it as .map files for selscan. 

#### pre-step. Generating map files. 

```
for i in {01..38}; do 
zcat $input/chr$i.ibgw.recode.vcf.gz | grep -v "^#" | awk '{print $1"\t"$3"\t"$2"\t"$2 }' > $input/chr$i.ibgw.map
done
```
#### Step a. Running selscan

```
selscan=/path_to_selscan/selscan
input=/path_to_phased_vcf
output=/path_to_output_selscan

for i in {01..38}; do
$selscan --xpnsl --vcf $input/chr$i.ibgw.recode.vcf.gz --vcf-ref $input/chr$i.negw.recode.vcf.gz --map $input/chr$i.ibgw.map --maf 0.05 --threads 12 --out $output/chr$i.ibgw --trunc-ok --max-extend-nsl -1

$selscan  --xpehh --vcf $input/chr$i.ibgw.recode.vcf.gz --vcf-ref $input/chr$i.negw.recode.vcf.gz --map $input/chr$i.ibgw.map --maf 0.05 --threads 12 --out $output/chr$i.ibgw --trunc-ok
done
```

#### Step b. Normalization

```
selscan=/path_to_selscan/norm

$norm --xpnsl --files $in/chr01.ibgw [...] $in/chr38.ibgw --winsize 100000 --bins 20
$norm --xpehh --files $in/chr01.ibgw [...] $in/chr38.ibgw --winsize 100000 --bins 20
```


# Detection of alleles under putative ancestry enrichment

Here, we are following Norris et al. (2018, BMC Genomics) strategy to calculate ancestry enrichment. 
We firstly estimate allele frequencies of DOG and NEGW as putative source populations for IBGW using VCFtools --freq2, taking into consideration the % of dog ancestry calculated by ELAI (4.93%)

```
in=/path_to_vcf
out=/path_to_output
vcftools=/path_to_vcftools/vcftools

for i in {01..38}; do 

for pop in dog ibgw negw; do
$vcftools --gzvcf $in/chr$i.$pop.recode.vcf.gz --freq2 --out $in/chr$i.$pop.freq2
done

chr$i.obsibgwref=$(cat $in/chr$i.ibgw.freq2.frq | tail -n+2 | awk '{ print $1, $2, $5}' | tr ' ' '\t')
chr$i.obsibgwalt=$(cat $in/chr$i.ibgw.freq2.frq | tail -n+2 | awk '{ print $1, $2, $6}' | tr ' ' '\t')
chr$i.dogref=$(cat $in/chr$i.dog.freq2.frq | tail -n+2 | awk '{ print $1, $2, $5 * 0.0493 }' | tr ' ' '\t')
chr$i.dogalt=$(cat $in/chr$i.dog.freq2.frq | tail -n+2 | awk '{ print $1, $2, $6 * 0.0493 }' | tr ' ' '\t')
chr$i.negwref=$(cat $in/chr$i.negw.freq2.frq | tail -n+2 | awk '{ print $1, $2, $5 * (1-0.0493) }' | tr ' ' '\t')
chr$i.negwalt=$(cat $in/chr$i.negw.freq2.frq | tail -n+2 | awk '{ print $1, $2, $6 * (1-0.0493) }' | tr ' ' '\t')
$in/chr$i.expibgwref=$(paste $dogref $negwref | awk '{ print $1, $2, $3 + $6 }'  | tr ' ' '\t')
$in/chr$i.expibgwalt=$(paste $dogalt $negwalt | awk '{ print $1, $2, $3 + $6 }'  | tr ' ' '\t')

for suffix in obsibgwref obsibgwalt expibgwref expibgwalt dogref dogalt negwref negwalt; do echo chr$i.$suffix >> autosomes.$suffix; done

done

# We estimate ancestry enrichment with a home script in python. This script will run a chi-square analysis and a post-hoc FDR correction as in Norris et al. (2018)

python anc_enrichment.py autosomes.dogalt autosomes.negwalt autosomes.expibgwref autosomes.expibgwalt autosomes.obsibgwref autosomes.obsibgwalt anc_enrich.table
```
